libs/s25main/WindowManager.cpp

```diff
@@ -876,3 +876,54 @@ void WindowManager::DrawToolTip()
         curTooltip->draw(ttPos);
     }
 }
+
+DrawPoint WindowManager::SnapWindow(Window* wndPtr, const Rect& wndRect) const
+{
+    const auto snapDistance = static_cast<int>(SETTINGS.interface.windowSnapDistance);
+    DrawPoint minSnapOffset = DrawPoint::all(0);
+    if(snapDistance == 0)
+        return minSnapOffset;
+
+    DrawPoint minDist(snapDistance + 1, snapDistance + 1);
+    auto calcSnapOffset = [](int a, int b, int& minDist, int& snapOffset) {
+        int diff = b - a;
+        int dist = std::abs(diff);
+        if(dist < minDist)
+        {
+            minDist = dist;
+            snapOffset = diff;
+        }
+    };
+
+    for(const auto& wnd : windows)
+    {
+        if(wnd.get() == wndPtr)
+            continue;
+
+        DrawPoint dist = DrawPoint::all(snapDistance + 1);
+        DrawPoint snapOffset = DrawPoint::all(0);
+        const auto calcSnapOffsetX = [&](int a, int b) { calcSnapOffset(a, b, dist.x, snapOffset.x); };
+        const auto calcSnapOffsetY = [&](int a, int b) { calcSnapOffset(a, b, dist.y, snapOffset.y); };
+        const Rect rect = wnd->GetBoundaryRect();
+
+        // Calculate smallest snap offset for each parallel pair of window edges
+        calcSnapOffsetX(wndRect.left, rect.left);
+        calcSnapOffsetX(wndRect.left, rect.right);
+        calcSnapOffsetX(wndRect.right, rect.left);
+        calcSnapOffsetX(wndRect.right, rect.right);
+        calcSnapOffsetY(wndRect.top, rect.top);
+        calcSnapOffsetY(wndRect.top, rect.bottom);
+        calcSnapOffsetY(wndRect.bottom, rect.top);
+        calcSnapOffsetY(wndRect.bottom, rect.bottom);
+
+        // Only snap along an axis if the windows overlap (+/- snap distance) along the orthogonal axis
+        if(dist.x < minDist.x && (rect.bottom + snapDistance) >= wndRect.top
+           && wndRect.bottom >= (rect.top - snapDistance))
+            minSnapOffset.x = snapOffset.x;
+        if(dist.y < minDist.y && (rect.right + snapDistance) >= wndRect.left
+           && wndRect.right >= (rect.left - snapDistance))
```

Comment:
I'd rewrite the overlap condition to be clearer by putting the same values on the same side:
```
        if((rect.right + snapDistance) >= wndRect.left
           && (rect.left - snapDistance) <= wndRect.right)
```

or at least for symmetry: 

```
        if(wndRect.left <= (rect.right + snapDistance)
           && rect.left <= (wndRect.right + snapDistance))
```

Reply:
I'm simply moving left-to-right and top-to-bottom within the conditions. This logical ordering made the most sense to me, but I'll take another look.